// manifest/dist/manifest/src/manifest/services/entity-manifest.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityManifestService = void 0;
const common_1 = require("@nestjs/common");
const types_1 = require("../../../../types/src");
const pluralize_1 = __importDefault(require("pluralize"));
const slugify_1 = __importDefault(require("slugify"));
const dasherize_1 = __importDefault(require("dasherize"));
const relationship_manifest_service_1 = require("./relationship-manifest.service");
const constants_1 = require("../../constants");
const manifest_service_1 = require("./manifest.service");
const hook_service_1 = require("../../hook/hook.service");
const policy_service_1 = require("../../policy/policy.service");
let EntityManifestService = class EntityManifestService {
    constructor(relationshipManifestService, manifestService, hookService, policyService) {
        this.relationshipManifestService = relationshipManifestService;
        this.manifestService = manifestService;
        this.hookService = hookService;
        this.policyService = policyService;
    }
    getEntityManifests(options) {
        const entities = Object.values(this.manifestService.getAppManifest({ fullVersion: true }).entities);
        if (!options?.fullVersion) {
            return entities.map((entity) => this.hideEntitySensitiveInformation(entity));
        }
        return entities;
    }
    getEntityManifest({ className, slug, fullVersion }) {
        if (!className && !slug) {
            throw new common_1.HttpException(`Either className or slug must be provided`, common_1.HttpStatus.BAD_REQUEST);
        }
        const entities = this.getEntityManifests({ fullVersion });
        let entityManifest;
        if (className) {
            entityManifest = entities.find((entity) => entity.className === className);
        }
        else {
            entityManifest = entities.find((entity) => entity.slug === slug);
        }
        if (!entityManifest) {
            throw new common_1.HttpException(`Entity ${className || slug} not found in manifest`, common_1.HttpStatus.NOT_FOUND);
        }
        if (!fullVersion) {
            return this.hideEntitySensitiveInformation(entityManifest);
        }
        return entityManifest;
    }
    transformEntityManifests(entitySchemaObject) {
        const entityManifests = Object.entries(entitySchemaObject).map(([className, entitySchema]) => {
            const partialEntityManifest = {
                className: entitySchema.className || className,
                nameSingular: entitySchema.nameSingular ||
                    pluralize_1.default.singular(entitySchema.className || className).toLowerCase(),
                slug: entitySchema.slug ||
                    (0, slugify_1.default)((0, dasherize_1.default)(entitySchema.single
                        ? entitySchema.className || className
                        : entitySchema.namePlural ||
                            pluralize_1.default.plural(entitySchema.className || className)).toLowerCase()),
                single: entitySchema.single || false,
                properties: (entitySchema.properties || [])
                    .filter((propSchema) => propSchema !== 'id' &&
                    propSchema.name !== 'id')
                    .map((propSchema) => this.transformProperty(propSchema, entitySchema)),
                hooks: this.transformHookObject(entitySchema.hooks),
                middlewares: entitySchema.middlewares || {}
            };
            if (entitySchema.single) {
                return this.getSingleEntityManifestProps(partialEntityManifest, entitySchema);
            }
            return this.getCollectionEntityManifestProps(partialEntityManifest, entitySchema);
        });
        entityManifests.forEach((entityManifest) => {
            if (entityManifest.single)
                return;
            entityManifest.relationships.push(...this.relationshipManifestService.getOneToManyRelationships(entityManifests, entityManifest));
        });
        entityManifests.forEach((entityManifest) => {
            if (entityManifest.single)
                return;
            entityManifest.relationships.push(...this.relationshipManifestService.getOppositeManyToManyRelationships(entityManifests, entityManifest));
        });
        return entityManifests;
    }
    getCollectionEntityManifestProps(partialEntityManifest, entitySchema) {
        if (entitySchema.authenticable) {
            partialEntityManifest.properties.push(...constants_1.AUTHENTICABLE_PROPS);
        }
        return {
            ...partialEntityManifest,
            properties: partialEntityManifest.properties,
            hooks: partialEntityManifest.hooks,
            namePlural: entitySchema.namePlural ||
                pluralize_1.default.plural(partialEntityManifest.className).toLowerCase(),
            mainProp: entitySchema.mainProp ||
                partialEntityManifest.properties.find((prop) => prop.type === types_1.PropType.String)?.name ||
                'id',
            seedCount: entitySchema.seedCount || constants_1.DEFAULT_SEED_COUNT,
            relationships: [
                ...(entitySchema.belongsTo || []).map((relationship) => this.relationshipManifestService.transformRelationship(relationship, 'many-to-one')),
                ...(entitySchema.belongsToMany || []).map((relationship) => this.relationshipManifestService.transformRelationship(relationship, 'many-to-many', partialEntityManifest.className))
            ],
            authenticable: entitySchema.authenticable || false,
            policies: {
                create: this.policyService.transformPolicies(entitySchema.policies?.create, constants_1.ADMIN_ACCESS_POLICY),
                read: this.policyService.transformPolicies(entitySchema.policies?.read, constants_1.ADMIN_ACCESS_POLICY),
                update: this.policyService.transformPolicies(entitySchema.policies?.update, constants_1.ADMIN_ACCESS_POLICY),
                delete: this.policyService.transformPolicies(entitySchema.policies?.delete, constants_1.ADMIN_ACCESS_POLICY),
                signup: entitySchema.authenticable
                    ? this.policyService.transformPolicies(entitySchema.policies?.signup, constants_1.PUBLIC_ACCESS_POLICY)
                    : [constants_1.FORBIDDEN_ACCESS_POLICY]
            }
        };
    }
    getSingleEntityManifestProps(partialEntityManifest, entitySchema) {
        return {
            ...partialEntityManifest,
            namePlural: partialEntityManifest.nameSingular,
            authenticable: false,
            mainProp: null,
            properties: partialEntityManifest.properties,
            hooks: partialEntityManifest.hooks,
            relationships: [],
            policies: {
                create: [constants_1.FORBIDDEN_ACCESS_POLICY],
                read: this.policyService.transformPolicies(entitySchema.policies?.read, constants_1.ADMIN_ACCESS_POLICY),
                update: this.policyService.transformPolicies(entitySchema.policies?.update, constants_1.ADMIN_ACCESS_POLICY),
                delete: [constants_1.FORBIDDEN_ACCESS_POLICY],
                signup: [constants_1.FORBIDDEN_ACCESS_POLICY]
            }
        };
    }
    transformProperty(propSchema, entitySchema) {
        if (typeof propSchema === 'string') {
            return {
                name: propSchema,
                type: types_1.PropType.String,
                hidden: false,
                validation: entitySchema.validation?.[propSchema] || {}
            };
        }
        return {
            name: propSchema.name,
            type: propSchema.type || types_1.PropType.String,
            hidden: propSchema.hidden || false,
            options: propSchema.options ||
                (propSchema.type === types_1.PropType.Image
                    ? { sizes: constants_1.DEFAULT_IMAGE_SIZES }
                    : {}),
            validation: Object.assign(entitySchema.validation?.[propSchema.name] ||
                {}, propSchema.validation),
            default: propSchema.default
        };
    }
    transformHookObject(hooksSchema) {
        return types_1.crudEventNames.reduce((acc, event) => {
            acc[event] = (hooksSchema?.[event] || []).map((hook) => this.hookService.transformHookSchemaIntoHookManifest(hook, event));
            return acc;
        }, {});
    }
    transformMiddlewareObject(middlewareSchema) {
        return types_1.crudEventNames.reduce((acc, event) => {
            acc[event] = (middlewareSchema?.[event] || []).map((middleware) => ({
                event,
                handler: middleware.handler
            }));
            return acc;
        }, {});
    }
    hideEntitySensitiveInformation(entityManifest) {
        return {
            ...entityManifest,
            properties: entityManifest.properties
                .filter((prop) => !prop.hidden)
                .map((prop) => {
                delete prop.hidden;
                return prop;
            })
        };
    }
};
exports.EntityManifestService = EntityManifestService;
exports.EntityManifestService = EntityManifestService = __decorate([
    (0, common_1.Injectable)(),
    __param(1, (0, common_1.Inject)((0, common_1.forwardRef)(() => manifest_service_1.ManifestService))),
    __metadata("design:paramtypes", [relationship_manifest_service_1.RelationshipManifestService,
        manifest_service_1.ManifestService,
        hook_service_1.HookService,
        policy_service_1.PolicyService])
], EntityManifestService);
