// manifest/dist/manifest/src/manifest/services/relationship-manifest.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelationshipManifestService = void 0;
const common_1 = require("@nestjs/common");
const common_2 = require("../../../../common/src");
const pluralize_1 = __importDefault(require("pluralize"));
let RelationshipManifestService = class RelationshipManifestService {
    transformRelationship(relationship, type, entityClassName) {
        if (type === 'many-to-one') {
            if (typeof relationship === 'string') {
                return {
                    name: (0, common_2.camelize)(relationship),
                    entity: relationship,
                    eager: false,
                    type
                };
            }
            return {
                name: (0, common_2.camelize)(relationship.name || relationship.entity),
                entity: relationship.entity,
                eager: relationship.eager || false,
                type
            };
        }
        else {
            if (typeof relationship === 'string') {
                return {
                    name: (0, pluralize_1.default)((0, common_2.camelize)(relationship)),
                    entity: relationship,
                    eager: false,
                    type,
                    owningSide: true,
                    inverseSide: (0, pluralize_1.default)((0, common_2.camelize)(entityClassName))
                };
            }
            return {
                name: (0, pluralize_1.default)((0, common_2.camelize)(relationship.name || relationship.entity)),
                entity: relationship.entity,
                eager: relationship.eager || false,
                type,
                owningSide: true,
                inverseSide: (0, pluralize_1.default)((0, common_2.camelize)(entityClassName))
            };
        }
    }
    getOneToManyRelationships(entityManifests, currentEntityManifest) {
        const oppositeRelationships = entityManifests
            .filter((otherEntityManifest) => otherEntityManifest.className !== currentEntityManifest.className)
            .reduce((acc, otherEntityManifest) => {
            const oppositeRelationship = otherEntityManifest.relationships.find((relationship) => relationship.entity === currentEntityManifest.className &&
                relationship.type === 'many-to-one');
            if (oppositeRelationship) {
                acc.push({
                    entity: otherEntityManifest,
                    relationship: oppositeRelationship
                });
            }
            return acc;
        }, []);
        return oppositeRelationships.map((oppositeRelationship) => {
            const relationship = {
                name: (0, common_2.camelize)(oppositeRelationship.entity.namePlural),
                entity: oppositeRelationship.entity.className,
                eager: false,
                type: 'one-to-many',
                inverseSide: oppositeRelationship.relationship.name
            };
            return relationship;
        });
    }
    getOppositeManyToManyRelationships(entityManifests, currentEntityManifest) {
        const oppositeRelationships = entityManifests
            .filter((otherEntityManifest) => otherEntityManifest.className !== currentEntityManifest.className)
            .reduce((acc, otherEntityManifest) => {
            const oppositeRelationship = otherEntityManifest.relationships.find((relationship) => relationship.entity === currentEntityManifest.className &&
                relationship.type === 'many-to-many' &&
                relationship.owningSide === true);
            if (oppositeRelationship) {
                acc.push({
                    entity: otherEntityManifest,
                    relationship: oppositeRelationship
                });
            }
            return acc;
        }, []);
        return oppositeRelationships.map((oppositeRelationship) => {
            const relationship = {
                name: (0, pluralize_1.default)((0, common_2.camelize)(oppositeRelationship.entity.namePlural)),
                entity: oppositeRelationship.entity.className,
                eager: false,
                type: 'many-to-many',
                owningSide: false,
                inverseSide: oppositeRelationship.relationship.name
            };
            return relationship;
        });
    }
};
exports.RelationshipManifestService = RelationshipManifestService;
exports.RelationshipManifestService = RelationshipManifestService = __decorate([
    (0, common_1.Injectable)()
], RelationshipManifestService);
