// manifest/dist/manifest/src/manifest/services/schema.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaService = void 0;
const common_1 = require("@nestjs/common");
const ajv_1 = __importDefault(require("ajv"));
const json_schema_1 = __importDefault(require("../../../../json-schema/src"));
const chalk_1 = __importDefault(require("chalk"));
let SchemaService = class SchemaService {
    validate(manifest) {
        this.validateAgainstSchema(manifest, json_schema_1.default[0]);
        this.validateCustomLogic(manifest);
        return true;
    }
    validateAgainstSchema(manifest, schema) {
        let validate = new ajv_1.default({
            schemas: json_schema_1.default,
            allowUnionTypes: true
        });
        validate = validate.getSchema(schema.$id);
        const valid = validate(manifest);
        if (!valid) {
            console.log(chalk_1.default.red('JSON Schema Validation failed. Please fix the following:'));
            validate.errors.forEach((error) => {
                console.log(chalk_1.default.red(JSON.stringify(error, null, 2)));
            });
            process.exit(1);
        }
        return true;
    }
    validateCustomLogic(manifest) {
        const entityNames = Object.keys(manifest.entities || {});
        Object.values(manifest.entities || {}).forEach((entity) => {
            const relationshipNames = Object.values(entity.belongsTo || []).map((relationship) => {
                if (typeof relationship === 'string') {
                    return relationship;
                }
                return relationship.entity;
            });
            relationshipNames.forEach((relationship) => {
                if (!entityNames.includes(relationship)) {
                    console.log(chalk_1.default.red('JSON Schema Validation failed. Please fix the following:'));
                    console.log(chalk_1.default.red(`Entity ${relationship} does not exist in the manifest`));
                    process.exit(1);
                }
            });
        });
        return true;
    }
};
exports.SchemaService = SchemaService;
exports.SchemaService = SchemaService = __decorate([
    (0, common_1.Injectable)()
], SchemaService);
