// manifest/dist/manifest/src/crud/controllers/collection.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollectionController = void 0;
const common_1 = require("@nestjs/common");
const crud_service_1 = require("../services/crud.service");
const auth_service_1 = require("../../auth/auth.service");
const policy_guard_1 = require("../../policy/policy.guard");
const rule_decorator_1 = require("../../policy/decorators/rule.decorator");
const is_collection_guard_1 = require("../guards/is-collection.guard");
const hook_interceptor_1 = require("../../hook/hook.interceptor");
const constants_1 = require("../../constants");
const middleware_interceptor_1 = require("../../middleware/middleware.interceptor");
const is_admin_guard_1 = require("../../auth/guards/is-admin.guard");
let CollectionController = class CollectionController {
    constructor(crudService, authService) {
        this.crudService = crudService;
        this.authService = authService;
    }
    async findAll(entitySlug, queryParams, req) {
        const isAdmin = await this.authService.isReqUserAdmin(req);
        return this.crudService.findAll({
            entitySlug,
            queryParams,
            fullVersion: isAdmin
        });
    }
    findSelectOptions(entitySlug, queryParams) {
        return this.crudService.findSelectOptions({
            entitySlug,
            queryParams
        });
    }
    async findOne(entitySlug, id, queryParams, req) {
        const isAdmin = await this.authService.isReqUserAdmin(req);
        return this.crudService.findOne({
            entitySlug,
            id,
            queryParams,
            fullVersion: isAdmin
        });
    }
    store(entity, entityDto) {
        return this.crudService.store(entity, entityDto);
    }
    put(entitySlug, id, itemDto) {
        return this.crudService.update({ entitySlug, id, itemDto });
    }
    patch(entitySlug, id, itemDto) {
        return this.crudService.update({
            entitySlug,
            id,
            itemDto,
            partialReplacement: true
        });
    }
    delete(entity, id) {
        return this.crudService.delete(entity, id);
    }
};
exports.CollectionController = CollectionController;
__decorate([
    (0, common_1.Get)('/:entity'),
    (0, rule_decorator_1.Rule)('read'),
    __param(0, (0, common_1.Param)('entity')),
    __param(1, (0, common_1.Query)()),
    __param(2, (0, common_1.Req)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object, Object]),
    __metadata("design:returntype", Promise)
], CollectionController.prototype, "findAll", null);
__decorate([
    (0, common_1.Get)(':entity/select-options'),
    (0, common_1.UseGuards)(is_admin_guard_1.IsAdminGuard),
    __param(0, (0, common_1.Param)('entity')),
    __param(1, (0, common_1.Query)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], CollectionController.prototype, "findSelectOptions", null);
__decorate([
    (0, common_1.Get)(':entity/:id'),
    (0, rule_decorator_1.Rule)('read'),
    __param(0, (0, common_1.Param)('entity')),
    __param(1, (0, common_1.Param)('id', common_1.ParseIntPipe)),
    __param(2, (0, common_1.Query)()),
    __param(3, (0, common_1.Req)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Number, Object, Object]),
    __metadata("design:returntype", Promise)
], CollectionController.prototype, "findOne", null);
__decorate([
    (0, common_1.Post)(':entity'),
    (0, rule_decorator_1.Rule)('create'),
    __param(0, (0, common_1.Param)('entity')),
    __param(1, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], CollectionController.prototype, "store", null);
__decorate([
    (0, common_1.Put)(':entity/:id'),
    (0, rule_decorator_1.Rule)('update'),
    __param(0, (0, common_1.Param)('entity')),
    __param(1, (0, common_1.Param)('id', common_1.ParseIntPipe)),
    __param(2, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Number, Object]),
    __metadata("design:returntype", Promise)
], CollectionController.prototype, "put", null);
__decorate([
    (0, common_1.Patch)(':entity/:id'),
    (0, rule_decorator_1.Rule)('update'),
    __param(0, (0, common_1.Param)('entity')),
    __param(1, (0, common_1.Param)('id', common_1.ParseIntPipe)),
    __param(2, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Number, Object]),
    __metadata("design:returntype", Promise)
], CollectionController.prototype, "patch", null);
__decorate([
    (0, common_1.Delete)(':entity/:id'),
    (0, rule_decorator_1.Rule)('delete'),
    __param(0, (0, common_1.Param)('entity')),
    __param(1, (0, common_1.Param)('id', common_1.ParseIntPipe)),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Number]),
    __metadata("design:returntype", Promise)
], CollectionController.prototype, "delete", null);
exports.CollectionController = CollectionController = __decorate([
    (0, common_1.Controller)(constants_1.COLLECTIONS_PATH),
    (0, common_1.UseGuards)(policy_guard_1.PolicyGuard, is_collection_guard_1.IsCollectionGuard),
    (0, common_1.UseInterceptors)(hook_interceptor_1.HookInterceptor, middleware_interceptor_1.MiddlewareInterceptor),
    __metadata("design:paramtypes", [crud_service_1.CrudService,
        auth_service_1.AuthService])
], CollectionController);
